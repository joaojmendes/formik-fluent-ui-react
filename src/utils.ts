import { FieldProps, getIn } from 'formik'

export type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>

export function getErrorMessage<T = any>({ field, form }: FieldProps<T>) {
  const error = getIn(form.errors, field.name)
  const touched = getIn(form.touched, field.name)

  return touched ? error : undefined
}

export function invokeAll(...callbacks: any[]) {
  return () => {
    for (const callback of callbacks) {
      if (callback && typeof callback === 'function') {
        callback()
      }
    }
  }
}

export function createFakeEvent({ name }: { name: string }) {
  return { target: { name } }
}
