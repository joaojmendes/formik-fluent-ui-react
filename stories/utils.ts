import { action } from '@storybook/addon-actions'
import { FormikActions } from 'formik'

export function handleSubmit<T>(
  values: T,
  { setSubmitting }: FormikActions<T>
) {
  action('submit:start')(values)
  setTimeout(() => {
    setSubmitting(false)
    action('submit:done')()
  }, 2000)
}
