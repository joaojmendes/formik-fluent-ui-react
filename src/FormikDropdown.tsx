import { Dropdown, IDropdownProps } from '@fluentui/react';
import { FieldProps } from 'formik'
import * as React from 'react'
import { createFakeEvent, getErrorMessage, invokeAll, Omit } from './utils'

export function mapFieldToDropdown<T = any>({
  form,
  field,
}: FieldProps<T>): Pick<
  IDropdownProps,
  'selectedKey' | 'selectedKeys' | 'onDismiss' | 'onChange' | 'errorMessage'
> {
  const multiSelect = Array.isArray(field.value)
  const shared = {
    errorMessage: getErrorMessage({ field, form }),
    onDismiss: () => field.onBlur(createFakeEvent(field)),
  }

  return multiSelect
    ? {
        ...shared,
        selectedKeys: field.value,
        onChange: (_, option) => {
          const value = field.value as any[]

          if (option!.selected) {
            form.setFieldValue(field.name, [...value, option!.key])
          } else {
            const idx = field.value.indexOf(option!.key)

            if (idx !== -1) {
              form.setFieldValue(field.name, [
                ...value.slice(0, idx),
                ...value.slice(idx + 1),
              ])
            }
          }
        },
      }
    : {
        ...shared,
        selectedKey: field.value,
        onChange: (_, option) => {
          form.setFieldValue(field.name, option!.key)
        },
      }
}

export type FormikDropdownProps<T> = Omit<IDropdownProps, 'selectedKey'> &
  FieldProps<T>
export function FormikDropdown<T = any>({
  field,
  form,
  ...props
}: FormikDropdownProps<T>) {
  const { errorMessage, onDismiss, ...fieldProps } = mapFieldToDropdown({
    field,
    form,
  })

  return (
    <Dropdown
      errorMessage={errorMessage}
      {...props}
      onDismiss={invokeAll(onDismiss, props.onDismiss)}
      {...fieldProps}
    />
  )
}
