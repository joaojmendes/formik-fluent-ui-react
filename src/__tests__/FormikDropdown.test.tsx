// tslint:disable:jsx-no-lambda

import { Field, FieldProps, Form, Formik } from 'formik'
import { Dropdown, setIconOptions } from '@fluentui/react'
import * as React from 'react'
import renderer from 'react-test-renderer'
import { FormikDropdown, mapFieldToDropdown } from '../FormikDropdown'
import { noop, serialize } from './utils'

// Suppress icon warnings.
setIconOptions({
  disableWarnings: true,
})

class Values {
  public single: null | string = 'foo'
  public multi: null | string[] = ['bar', 'baz']
}

function createFieldProps(value: null | string | string[]): FieldProps<null | string | string[]> {
  return {
    field: {
      value,
      onChange: jest.fn(),
      onBlur: jest.fn(),
      name: Array.isArray(value) ? 'multi' : 'single',
    },
    form: { setFieldValue: jest.fn(), handleBlur: jest.fn(() => jest.fn()) },
  } as any
}

const options = ['foo', 'bar', 'baz'].map(x => ({ key: x, text: x }))

test('<FormikDropdown /> renders correctly as a field component', () => {
  const component = renderer.create(
    <Formik initialValues={new Values()} onSubmit={noop}>
      <Form>
        <Field
          name="single"
          render={(fieldProps: FieldProps<any>) => (
            <FormikDropdown {...fieldProps} options={options} />
          )}
        />
        <Field
          name="multi"
          render={(fieldProps: FieldProps<any>) => (
            <FormikDropdown {...fieldProps} options={options} />
          )}
        />
      </Form>
    </Formik>
  )

  expect(component.toJSON()).toMatchSnapshot()
})

test('<FormikDropdown /> renders a Fabric <Dropdown />', () => {
  const fieldProps = createFieldProps('foo')

  const formikDropdown = renderer.create(
    <FormikDropdown {...fieldProps} options={options} />
  )
  const fabricDropdown = renderer.create(
    <Dropdown {...mapFieldToDropdown(fieldProps)} options={options} />
  )
  expect(serialize(formikDropdown)).toBe(serialize(fabricDropdown))
})

test('mapFieldToDropdown() maps FieldProps to IDropdownProps for single select', () => {
  const { field, form, meta } = createFieldProps('foo')

  const props = mapFieldToDropdown({ form, field, meta })

  expect(props.selectedKey).toBe(field.value)

  props.onChange!(null as any, options[1])

  expect(form.setFieldValue).toHaveBeenCalledTimes(1)
  expect(form.setFieldValue).toHaveBeenCalledWith(field.name, options[1].key)
})

test('mapFieldToDropdown() maps FieldProps to IDropdownProps for multi select', () => {
  const { field, form, meta } = createFieldProps(['bar', 'baz'])

  const props = mapFieldToDropdown({ form, field, meta })

  expect(props.selectedKeys).toBe(field.value)

  // remove 'bar'
  props.onChange!(null as any, { ...options[1], selected: false })

  expect(form.setFieldValue).toHaveBeenCalledTimes(1)
  expect(form.setFieldValue).toHaveBeenCalledWith(field.name, ['baz'])

  // add 'foo
  props.onChange!(null as any, { ...options[0], selected: true })

  expect(form.setFieldValue).toHaveBeenCalledTimes(2)
  expect(form.setFieldValue).toHaveBeenCalledWith(field.name, [
    'bar',
    'baz',
    'foo',
  ])
})
