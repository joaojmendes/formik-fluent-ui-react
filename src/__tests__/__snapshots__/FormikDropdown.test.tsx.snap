// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormikDropdown /> renders correctly as a field component 1`] = `
<form
  onReset={[Function]}
  onSubmit={[Function]}
>
  <div
    className="ms-Dropdown-container root-47"
  >
    <div
      aria-activedescendant="Dropdown1-option"
      aria-describedby="Dropdown1-option"
      aria-expanded="false"
      className="ms-Dropdown dropdown-49"
      data-is-focusable={true}
      id="Dropdown1"
      onBlur={[Function]}
      onChange={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      role="listbox"
      tabIndex={0}
    >
      <span
        aria-atomic={true}
        aria-label="foo"
        aria-live="off"
        aria-posinset={1}
        aria-selected={true}
        aria-setsize={3}
        className="ms-Dropdown-title title-50"
        id="Dropdown1-option"
        role="option"
      >
        <span>
          foo
        </span>
      </span>
      <span
        className="ms-Dropdown-caretDownWrapper caretDownWrapper-51"
      >
        <i
          aria-hidden={true}
          className="ms-Dropdown-caretDown caretDown-65"
          data-icon-name="ChevronDown"
          role="presentation"
        />
      </span>
    </div>
  </div>
  <div
    className="ms-Dropdown-container root-47"
  >
    <div
      aria-activedescendant="Dropdown2-option"
      aria-describedby="Dropdown2-option"
      aria-expanded="false"
      className="ms-Dropdown dropdown-49"
      data-is-focusable={true}
      id="Dropdown2"
      onBlur={[Function]}
      onChange={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      role="listbox"
      tabIndex={0}
    >
      <span
        aria-atomic={true}
        aria-live="off"
        aria-setsize={3}
        className="ms-Dropdown-title ms-Dropdown-titleIsPlaceHolder title-66"
        id="Dropdown2-option"
        role="option"
      />
      <span
        className="ms-Dropdown-caretDownWrapper caretDownWrapper-51"
      >
        <i
          aria-hidden={true}
          className="ms-Dropdown-caretDown caretDown-65"
          data-icon-name="ChevronDown"
          role="presentation"
        />
      </span>
    </div>
  </div>
</form>
`;
